drivers: flash: merger Generic SPI Flash driver

findlayfeng opened the pull request:
Merged #5456.
Update spi api.
Add the CS/gpio logic.
Independent of spi flash config.

fabled committed:
driver: flash: add a Generic SPI Flash driver

This drivers supports chips with Serial Flash Discoverable
Parameters (SFDP) information.

Add Kconfig.spi
Add spi_flash_sfdp.c
Add spi_flash_sfdp.h
Support spi dual mode
Partial support spi quad mode

Signed-off-by: Timo Teräs <timo.teras@iki.fi>
Signed-off-by: Findlay Feng <i@fengch.me>

findlayfeng commented:
#7532

tbursztyka reviewed (approved):
Seems fine to me (remove the 'etc' in last patch's commit title).

Note that you don't have to close and create a new PR each time you make changes to your patches: a git push -f will just do, and all will be updated in existing PR.

findlayfeng commented:
嗯，我尝试过`push -f`但是，不清楚是因为有三次commit 还是因为是一个空格改制表符的空提交，一直成功但没有被修改。正好提示有一个不兼容的合并，于是我就新拉取了一个分支做处理。
Below from Google Translate
Well, I tried `push -f` , and it's unclear whether there were three commit or because it was an empty commit of a space change tab, it was successful but not modified. Just in case there is an incompatible merge, so I took a new branch to do the processing.

findlayfeng commented:
@nashif 
一个奇怪的错误，看起来不是我的问题
Below from Google Translate
A strange error that does not seem to be my problem

findlayfeng commented:
@tbursztyka Made more changes in this patch

tbursztyka commented on drivers/flash/Kconfig:
```diff
@@ -36,6 +36,39 @@ config FLASH_PAGE_LAYOUT
+	default ""
+
+config SPI_FLASH_SFDP_INIT_PRIORITY
+	int
```
no prompt for that one?

tbursztyka commented on drivers/flash/Kconfig:
```diff
@@ -36,6 +36,39 @@ config FLASH_PAGE_LAYOUT
+	  Flash Discoverable Parameters (SFDP).
+
+config SPI_FLASH_SFDP_SPI_NAME
+	string
```
get rid of prompt and put the prompt message here.

tbursztyka commented on drivers/flash/spi_flash_sfdp.c:
```diff
@@ -0,0 +1,558 @@
+				phdr[i].major_ver, phdr[i].minor_ver,
+				get_phdr_addr(&phdr[i]),
+				phdr[i].length);
+		switch (get_phdr_id(&phdr[i])) {
```
why a switch if there is only 1 case?

tbursztyka commented on drivers/flash/spi_flash_sfdp.c:
```diff
@@ -0,0 +1,558 @@
+	wait_for_flash_idle(dev);
+	i = 0;
+	hdr[i++] = data->read_opcode;
+	if (data->addr_len == 4) {
```
add an empty line before

tbursztyka commented on drivers/flash/spi_flash_sfdp.c:
```diff
@@ -0,0 +1,558 @@
+	if (data->addr_len == 4) {
+		hdr[i++] = offset >> 24;
+	}
+	hdr[i++] = offset >> 16;
```
same

tbursztyka commented on drivers/flash/spi_flash_sfdp.c:
```diff
@@ -0,0 +1,558 @@
+	hdr[i++] = offset >> 16;
+	hdr[i++] = offset >> 8;
+	hdr[i++] = offset;
+	r = spi_transceive(&data->config, txbufs, ARRAY_SIZE(txbufs),
```
same

tbursztyka commented on drivers/flash/spi_flash_sfdp.c:
```diff
@@ -0,0 +1,558 @@
+
+	SYS_LOG_DBG("write: @%x len %x, opcode %x (page size %d)",
+		offset, len,
+		data->write_opcode, data->page_size);
```
hum, you can put that on previous line

tbursztyka commented on drivers/flash/spi_flash_sfdp.c:
```diff
@@ -0,0 +1,558 @@
+
+	i = 0;
+	hdr[i++] = data->write_opcode;
+	if (data->addr_len == 4) {
```
add an empty line before (basicall: if assigned value is not used in the if condition: break a line).

tbursztyka commented on drivers/flash/spi_flash_sfdp.c:
```diff
@@ -0,0 +1,558 @@
+	if (data->addr_len == 4) {
+		hdr[i++] = offset >> 24;
+	}
+	hdr[i++] = offset >> 16;
```
same here (break a line if it's not another '}' etc...)

tbursztyka reviewed (changes requested):
you will need to rebase and update that against latest SPI API. Also, add support for CS GPIO (see existing code such as spi_flash_w25qxxdv.c)

findlayfeng commented:
commit 88688ce  来自 Merged #5456，我只是让它变得可合并而已,我在 commit ff2e7ec 中做了自己的工作，包括gpio_cs，几乎重写了。
From google translate
Commit 88688ce from Merged #5456, I just made it mergeable, I did my work in commit ff2e7ec, including gpio_cs, which was almost rewritten.
@tbursztyka

Commit:
subsys: flash_map: Add SPI_FLASH to flash_drivers_map

id SPI_FLASH_0_ID;
name CONFIG_SPI_FLASH_DRV_NAME;

Signed-off-by: Findlay Feng <i@fengch.me>

tbursztyka commented on drivers/flash/spi_flash_sfdp.h:
```diff
@@ -0,0 +1,290 @@
+#define CONFIG_SPI_FLASH_SMPT_SIZE 16
+#endif
+
+typedef u32_t dword_t;
```
why this typedef? Seems unnecessary

tbursztyka commented on drivers/flash/spi_flash_sfdp.c:
```diff
@@ -1,558 +1,993 @@
 /*
  * Copyright (c) 2017 Crypta Labs Ltd
+ * Copyright (c) 2018 Findlay Feng
```
Merge this specific modification of sfdp to your first commit.

And this actual commit becomes one for adding microchip table support then.

findlayfeng replied:
I will change it

tbursztyka commented on drivers/flash/spi_flash_sfdp.c:
```diff
@@ -1,558 +1,993 @@
-		.frequency = CONFIG_SPI_FLASH_SFDP_SPI_FREQ_0,
-		.operation = SPI_WORD_SET(8),
-	};
+#if defined(SPI_FLASH_GPIO_SPI_CS)
```
Why not a Kconfig option for this? So it would be CONFIG_ prefixed then.

findlayfeng replied:
oh...
This is a lapse

Commit:
driver: spi_flash: add Microchip SFDP table

Supports Microchip SFDP table

Add Kconfig.spi.microchip
Add spi_flash_sfdp_microchip.h
Add spi_flash_sfdp_microchip.c

Signed-off-by: Findlay Feng <i@fengch.me>

fabled commented:
Also overriding the configuration SPI freq does not seem right. This happens for reading SFDP table, and when microchip flash is detected.

findlayfeng commented:
I only tried the standard spi, it is estimated that the dual spi is working, the quad spi still needs some extra work, I have left a lot of todo, and can improve it when needed.
Currently I don't have the right hardware to test.

findlayfeng commented:
There is a definition of the read frequency that should be used in the SFDP document. This frequency is not necessarily the same as the one specified in the chip documentation.

findlayfeng commented:
I use the nrf5x spi driver, the memory address of the configuration information does not change without modifying the actual configuration of the spi. . .

fabled commented:
It is still normal that lower frequency is needed to be used. I think the configuration provided frequency should be honored. I rebased my spi sfdp driver on top of current head just to get everything I needed working - e.g. seems the microchip global unlock is not yet supported properly.

findlayfeng commented:
page 3
```
4.4 Clock Rate
SFDP compliant devices must support 50 MHz operation for the Read SFDP command (instruction 5Ah).
Devices may support a wider frequency range, but a controller can always run SFDP cycles at 50 MHz or
less and get valid results. 
```

galak commented:
We have a spi_nor driver in tree.  Please update any features like SFDP to that driver.  Closing this PR at this point.
